### DISCUSSION TITLE
Realtime capabilities of madmom

### DISCUSSION THREAD

**Discussion by space-mob** [CPJKU/madmom#493] [General]
Hi @superbock
first and foremost merry xmas 🎅 and thank you for all the work you put into this library. 
Over the last few days i've been doing some research regarding audio feature extraction and music information retrieval. 
I'm especially interested in applications where future data is unknown, i.e., real-time / streamed processing.

I've gone through opened/closed issues and the old discussions and found several related threads:
- [DBNBeatTracker on Stream Input](https://groups.google.com/g/madmom-users/c/yYtp6Y43yWQ/m/OcigMLSMBgAJ)
- [online onset RNN detection](https://groups.google.com/g/madmom-users/c/pF5aaLsvFAs/m/BI3FJrvyBQAJ)
- [process audio input from microphone?](https://groups.google.com/g/madmom-users/c/jVA4XYgakfI/m/AfKQ5lmzCAAJ)
- [Documentation/Examples for online mode/real-time processing](https://github.com/CPJKU/madmom/issues/484)
- [Real time beat detection](https://github.com/CPJKU/madmom/issues/382)
- etc.

From my understanding you have already integrated pyAudio in some of the processors, most of them neural network based. 
To cut to the chase - i'm especially curious if the implementation based on your [SuperFlux](http://phenicx.upf.edu/system/files/publications/Boeck_DAFx-13.pdf) paper already works to extract an onset envelopes in real-time? 

If not, what would need to be done to make it work? I assume we would need an additional layer to the pyAudio stream that would maybe have some kind of rolling buffer and handle the overlap?

best regards,
david

**Comment by superbock** [comment]
Hi David,

Sorry for the late reply!

No SuperFlux is not online ready yet. However, it is not too complicated to achieve this. The only missing part is to cache the previous spectrogram frame to be able to compute the difference of the current frame with it. If you are still interested in doing this I can give you more detailed hints on how to accomplish this.

**Reply by un1tz3r0** [reply]
Hi I've sat down several times with the intention of extending madmom to do just this, but alas it's not happened yet. Any pointers you could give, @superbock, would be a huge help.

I've been working on a python binding for pipewire with the intention of doing super-low-latency analysis with madmom.
